app/foi/models.py

```diff
@@ -0,0 +1,135 @@
+
+
+class FoiIndexPage(BasePageWithRequiredIntro):
+    class Meta:
+        verbose_name = "Freedom of information listing page"
+
+    max_count = 1
+    subpage_types = ["foi.FoiRequestPage"]
+
+
+class FoiRequestPage(BasePage):
+    class Meta:
+        verbose_name = "Freedom of information request page"
+        verbose_name_plural = "Freedom of information request pages"
+
+    parent_page_types = ["foi.FoiIndexPage"]
+    subpage_types = []
+
+    # Not a required field like in BasePage
+    teaser_text = models.TextField(
+        verbose_name="teaser text",
+        help_text="A short, enticing description of this page. This will appear in promos and under thumbnails around the site.",
+        max_length=160,
+        null=True,
+        blank=True,
+    )
+
+    reference = models.CharField(
+        verbose_name="request reference",
+        max_length=100,
+        null=True,
+    )
+
+    date = models.DateField(
+        verbose_name="publication date",
+        help_text="The date the request was published.",
+        null=True,
+    )
+
+    request = StreamField(
+        RequestStreamBlock,
+        null=True,
+    )
+
+    REQUEST_OUTCOMES = {
+        "P": "Information provided",
+        "S": "Some information provided",
+        "N": "Information not held",
+        "W": "Information withheld",
+        "R": "Request refused",
+    }
+    outcome = models.CharField(
+        max_length=1,
+        choices=REQUEST_OUTCOMES,
+    )
+
+    def outcome_description(self):
+        return self.REQUEST_OUTCOMES.get(self.outcome, "")
+
+    response = StreamField(
+        ResponseStreamBlock,
+        null=True,
+    )
+
+    annexe = StreamField(
+        AnnexeStreamBlock,
+        null=True,
+        blank=True,
+    )
+
+    @cached_property
+    def type_label(cls) -> str:
+        return "Freedom of information request"
+
+    content_panels = BasePage.content_panels + [
+        MultiFieldPanel(
+            heading="Request",
+            classname="collapsible",
+            children=[
+                FieldPanel("reference"),
+                FieldPanel("date"),
+                FieldPanel("request"),
+            ],
+        ),
+        FieldPanel("outcome"),
+        FieldPanel("response"),
+        FieldPanel("annexe"),
+    ]
+
+    search_fields = BasePage.search_fields + [
+        index.SearchField("reference"),
+        index.SearchField("request"),
+        index.SearchField("outcome_description"),
+        index.SearchField("response"),
+        index.SearchField("annexe"),
+    ]
+
+    api_fields = BasePage.api_fields + [
+        APIField("reference"),
+        APIField("date"),
+        APIField("request"),
+        APIField("outcome_description"),
+        APIField("response"),
+        APIField("annexe"),
+    ]
+
+    def clean(self, *args, **kwargs):
+        if self.reference:
+            slug = slugify(self.reference)
+            if slug not in self.slug:
```

Comment:
It might be better to do `if slug != self.slug` because I had a scenario where I put in a reference "TEST1234" but then updated to "TEST123" and as "TEST123" is in "TEST1234", the slug wasn't updated - so remained "TEST1234" for a page of reference "TEST123".

Reply:
Didn't consider the redirects - good point, but if the reference changes we would be creating one as well, wouldn't we? If I go from TEST123 to TEST1234 then I should update the slug to `test1234` - or no?